### ISSUE TITLE
Support command line -- to stop parsing options and start parsing input filenames

### ISSUE THREAD

**Issue by weswigham** [Microsoft/TypeScript#9092]
**TypeScript Version:** 

1.7.5 / 1.8.0-beta / nightly (1.9.0-dev.20160217)

**Expected behavior:**

You should be able to write `tsc --sourcemap --noEmitHelpers -- helpers.ts fileA.ts fileB.ts`, where `--` terminates compiler options parsing.

**Actual behavior:** 

`--` is not recognized as a terminator for command line options, instead we throw `error TS5023: Unknown compiler option ''.`.

**Comment by sandersn** [comment]
This is the way that git makes sure that the boundary between options and files is not ambiguous.

**Comment by coreh** [comment]
It's not only a git thing, all the UNIX apps that use `getopt` [inherit that behavior](http://unix.stackexchange.com/questions/147143/when-and-how-was-the-double-dash-introduced-as-an-end-of-options-delimiter).

**Comment by aluanhaddad** [comment]
We don't need another way to pass command line arguments just leave it as it is please.

**Comment by billyzkid** [comment]
I'm surprised this was closed since it is standard behavior supported by just about every other commonly used npm package/script/cli, e.g. prettier, eslint, git, npm, etc.

Here's at least one problem it solves. I don't have an easy way to compile files from the command line that start with "-" or "--" in their name. For example, if I execute `eslint --version.ts` or `prettier --version.ts`, then I either get the version output (from eslint) or an error message (from prettier). However, in both cases, I can easily fix this by altering my commands slightly to `eslint -- --version.ts` and `prettier -- --version.ts`.

Unfortunately, I can't easily fix this with `tsc` as shown below.

```
PS C:\repos\monorepo\packages\a> npm run compile

> @monorepo/package-a@1.0.0 compile
> tsc --version.ts

error TS5025: Unknown compiler option '--version.ts'. Did you mean 'version'?
PS C:\repos\monorepo\packages\a> npm run compile

> @monorepo/package-a@1.0.0 compile
> tsc -- --version.ts

error TS5023: Unknown compiler option '--'.
error TS5025: Unknown compiler option '--version.ts'. Did you mean 'version'?
```

Now using eslint. Note the command initially fails, then succeeds if I add the "--".

```
PS C:\repos\monorepo\packages\a> npm run lint

> @monorepo/package-a@1.0.0 lint
> eslint --version.ts

v8.45.0
PS C:\repos\monorepo\packages\a> npm run lint

> eslint -- --version.ts
```

**Comment by RyanCavanaugh** [comment]
Why do you have files named that way?

**Comment by coreh** [comment]
> Why do you have files named that way?

You really shouldn't, at least normally.

This specific command line parsing behavior is there mostly as a robustness/command injection prevention mechanism for shell scripting. It's (very) uncommon to have files starting with a dash, but if you have a shell script with, say, a glob and accidentally _do have_ a file with a dash present, it can mess things up in ways that are hard to understand and debug. (And potentially even be a security vulnerability.)

Now, invoking `tsc` in a shell script with user generated content outside of a sandbox is a very reckless thing to do, and likely not a “supported” use-case, so I don't think the security argument is very important in this case. The main reason why I think this would be a nice addition is consistency with UNIX tools: Native commands that use `getopt` get this for free, and Node.js CLIs that use the most popular command line parsing libraries also inherit this behavior. (Hence why `prettier` and `eslint` support it.)

If you've been bit in the past by files being accidentally parsed as options, you'll defensively add this to all your shell scripts when dealing with globs, environment variables, or the expanded output of other shell commands. ([It's very common!](https://github.com/search?q=%22+--+%22+language%3AShell+&type=code&p=4)) You'll then find that `tsc` doesn't support it, and have to update your script to remove it, which is a _tiny_ paper cut, but still a paper cut. (This is why I liked /commented on this issue back in 2016.)

---

**Some additional context:** Some commands like `git` and `npm` give _additional_ meaning to the `--`, besides just not parsing the options. On `git` it disambiguates between branch and file names, and on `npm` it forwards the options to the script being invoked by `npm run`. I don't think either applies to `tsc`.

**Comment by billyzkid** [comment]
> Why do you have files named that way?

Yes, it's extremely unusual to name files this way intentionally, but it's still legal to do so. And if you are executing `tsc` programmatically and/or specifying the arguments dynamically as I need to do, using "--" is a safe way to ensure the arguments that I'm passing in are never misinterpreted.
